src/coreclr/jit/gentree.cpp

```diff
@@ -25404,8 +25391,620 @@ GenTree* Compiler::gtNewSimdRoundNode(var_types type, GenTree* op1, CorInfoType
     return gtNewSimdHWIntrinsicNode(type, op1, intrinsic, simdBaseJitType, simdSize);
 }
 
+GenTree* Compiler::gtNewSimdShuffleNodeVariable(
+    var_types type, GenTree* op1, GenTree* op2, CorInfoType simdBaseJitType, unsigned simdSize, bool isShuffleNative)
+{
+    assert(IsBaselineSimdIsaSupportedDebugOnly());
+
+    assert(varTypeIsSIMD(type));
+    assert(getSIMDTypeForSize(simdSize) == type);
+
+    assert(op1 != nullptr);
+    assert(op1->TypeIs(type));
+
+    var_types simdBaseType = JitType2PreciseVarType(simdBaseJitType);
+    assert(op2 != nullptr);
+    assert(op2->TypeIs(type));
+    assert(!op2->IsCnsVec() || isShuffleNative);
+
+    GenTree* retNode = nullptr;
+    GenTree* cnsNode = nullptr;
+
+    size_t elementSize  = genTypeSize(simdBaseType);
+    size_t elementCount = simdSize / elementSize;
+
+    // duplicate operand 2 for non-ShuffleNative implementation later
+    // (also, on arm64, byte ShuffleNative is same as Shuffle)
+    GenTree* op2DupSafe = nullptr;
+#if defined(TARGET_XARCH)
+    if (!isShuffleNative)
+#elif defined(TARGET_ARM64)
+    if (!isShuffleNative && elementSize > 1)
+#else
+#error Unsupported platform
+#endif // !TARGET_XARCH && !TARGET_ARM64
+    {
+        op2DupSafe = fgMakeMultiUse(&op2);
+    }
+
+#if defined(TARGET_XARCH)
+    // on xarch, signed comparison is cheaper, so whenever we are able to use it in the
+    // result & (indices < elementCount) step for Shuffle, we do. Specifically, we are
+    // able to use it when the top bit causes zeroing (then we can compare indices as
+    // if they were signed, since negative cases are already handled).
+    bool canUseSignedComparisonHint = false;
+
+    // TODO-XARCH-CQ: If we have known set/unset bits for the indices, we could further optimise many cases
+    // below.
+
+    bool isV512Supported = false;
+    if (simdSize == 64)
+    {
+        if (elementSize == 1)
+        {
+            assert(compIsaSupportedDebugOnly(InstructionSet_AVX512VBMI));
+
+            // swap the operands to match the encoding requirements
+            retNode = gtNewSimdHWIntrinsicNode(type, op2, op1, NI_AVX512VBMI_PermuteVar64x8, simdBaseJitType, simdSize);
```

Comment:
I don't see where and how you handle op1 & op2 side effect re-ordering

Reply:
It's handled on the next line: `retNode->SetReverseOp();`

Reply:
> It's handled on the next line: `retNode->SetReverseOp();`

My understanding that you accept a HW_INTRINSIC(op1, op2) and then lower down like this. If HW_INTRINSIC nodes support "reverse ops" - where do you check that the source HW_INTRINSIC wasn't already marked as ReversedOp ?

I think the ReverseOp is a dangerous concept that we should remove

Reply:
I guess you can leave it as is, but seems like you ignore the fact the operands already came in a reversed order

Reply:
> My understanding that you accept a HW_INTRINSIC(op1, op2) and then lower down like this. If HW_INTRINSIC nodes support "reverse ops" - where do you check that the source HW_INTRINSIC wasn't already marked as ReversedOp ?

I created it on the line you commented on, directly above the `retNode->SetReverseOp();` - I can add an assertion checking it's not already reversed if you like?

Reply:
that is not what I meant, the function that you wrote is processing an existing `HW_INTRINSIC(op1, op2)` node, am I correct? In the rationalizer. you take those op1 and op2 and compose a new tree

Reply:
> I think the ReverseOp is a dangerous concept that we should remove, opinions @dotnet/jit-contrib ?

Yeah, number of places in the JIT do not have sufficient handling of `GTF_REVERSE_OPS`, so introducing new places that set it is worrying. I agree that it looks like the new code added in rationalization is not properly handling the possibility of `GTF_REVERSE_OPS` as well.

Reply:
I believe this has been fixed correctly now, as discussed on discord - please re-open if not (same with other ones)